### ISSUE TITLE
Weight averaging across checkpoints (SWA): we keep the last iterate and throw the trajectory away

### ISSUE THREAD

**Issue by nleck** [stSoftwareAU/NEAT-TS#3918]
## What is missing

Backpropagation in NEAT-AI ends by keeping the weights it happened to hold at
the final iteration. There is no averaging of the weights visited along the way.

Stochastic Weight Averaging — Izmailov, Podoprikhin, Garipov, Vetrov & Wilson
(2018), *Averaging Weights Leads to Wider Optima and Better Generalization* — is
one line of extra state and reliably lands in a flatter, better-generalising
region than the final iterate:

```
w_SWA  =  (1/n) Σ w_i        over checkpoints i taken late in training
```

That is the whole method. No extra forward passes, no extra gradients — one
running mean, updated every `k` iterations after a warm-up, and the average is
what you keep.

## Why it fits NEAT-AI specifically

**We already record weight trajectories.** Memetic evolution preserves a
creature's pre-mutation state and records weight/bias deltas after training
(`src/blackbox/MemeticTrajectory.ts`). The information SWA needs is being
collected already for a different purpose; SWA is a second, cheaper consumer of
it.

**The topology is fixed during a backprop run**, which is the precondition — the
average of two weight vectors is only meaningful when they index the same
network. Within one `trainDir`-style epoch loop that holds. Across a structural
mutation it does not, and the implementation must refuse to average across one
rather than silently produce nonsense.

**Flat optima are the thing we currently cannot detect.** Accepted improvements
on the production creature run around 6e-05, and there is no way today to tell a
robust win from one that a small weight perturbation would erase (tracked as
stSoftwareAU/NEAT-AI-scorer#588). SWA attacks the same problem from the training
side rather than the acceptance side: instead of *detecting* sharp optima, it
avoids landing in them. The two are complementary, and #588 is the natural way
to measure whether this one worked.

**It composes with the learning-rate schedules already present.**
`learningRateStrategy: "warm_restart"` (`src/propagate/BackPropagation.ts:97`)
is the cyclic schedule the SWA paper pairs with — each restart visits a
different point in the same basin, which is precisely the set of iterates worth
averaging. The pieces are already here and unconnected.

## Proposed

Add to `BackPropagationArguments`:

| Option | Default | Meaning |
| --- | --- | --- |
| `weightAveraging` | `"none"` | `"none"` or `"swa"` |
| `swaStartFraction` | `0.75` | begin averaging after this fraction of the run — early iterates are noise and must not enter the mean |
| `swaFrequency` | every iteration | capture interval |

Behaviour:

- Maintain a running mean of weights **and biases** — a single extra vector, not
  a list of checkpoints.
- Refuse to average across a topology change; assert on it rather than tolerate it.
- The averaged weights **replace** the final iterate only if they score better.
  Everything in this fleet is decided by the scorer, and this should be no
  different — SWA is normally applied unconditionally, but we have an
  authoritative judge and no reason not to use it.
- Averaging state is per-run and must **not** enter the exported creature.

## The one real caveat

SWA in its published form assumes batch-normalisation statistics are recomputed
after averaging. NEAT-AI has no batch norm, which makes this simpler rather than
harder — there is no post-averaging pass to run. Worth stating in the docs so
nobody later "fixes" its absence.

## How to know whether it worked

The honest test is not in-sample error. Averaging is a *generalisation*
intervention, so an in-sample improvement is a weak signal and an in-sample
regression is not necessarily a failure. Measure it against
stSoftwareAU/NEAT-AI-scorer#588 once that lands: does an SWA-trained creature
survive weight perturbation better than the final-iterate one at equal score?
That is the question SWA is designed to answer.

## Acceptance

- [ ] `weightAveraging: "swa"` implemented with a running mean over weights and biases.
- [ ] Averaging refuses to span a topology change, with a test.
- [ ] Averaged weights adopted only on a scorer-verified improvement.
- [ ] Averaging state excluded from creature export, with a test.
- [ ] Documented interaction with each `learningRateStrategy`, `"warm_restart"` in particular.
- [ ] A measurement on the production creature, reported whichever way it goes.

## References

- Izmailov et al. (2018), *Averaging Weights Leads to Wider Optima and Better Generalization*.
- Polyak & Juditsky (1992), *Acceleration of Stochastic Approximation by Averaging* — the original iterate-averaging result.
- Garipov et al. (2018), *Loss Surfaces, Mode Connectivity, and Fast Ensembling of DNNs* — why the averaged point is a good one.
